iceberg-rust/Cargo.toml

```diff
@@ -30,4 +30,5 @@ thrift = { version = "0.17.0", default-features = false }
 thiserror = { workspace = true }
 derive-getters = { workspace = true }
 iceberg-rust-spec = { path = "../iceberg-rust-spec", version = "0.5.8" }
+smallvec = { version = "1.13.2", features = ["const_generics"] }
```

Comment:
The type alias is a great idea!

Generally, I don't know the number of partition columns beforehand and I cannot create the according data-structure at compile time. That's why I might need to allocate memory at runtime.
However, the number of partition columns is typically very small, that's why I would like to reserve the according memory on the stack and only spill to the heap if there are more partition columns. That's what smallvec does. It tries to do everything in an array on the stack and allocates additional memory if needed. This way the common case should be faster, but nothing happens in case there are more columns.

Reply:
I agree the optimization makes sense, but I was wondering if it is really necessary? Is the allocation really slowing down the execution? I think this is only the case if you make a huge amount of allocations. We can assume here that it's the case, but I was curious how you got to that conclusion.

When you create abstractions to encapsulate part of your code, you generally want to keep only the minimal information in the signatures to keep them as readable as possible. Here, having a specific `Vec` type adds noise. Aliasing to `Vec4` doesn't really help, because the `4` is an optimization value.

Reply:
I would be tempted to have a type like `PartitionVec` that is used across the entire crate. That would semantically indicate that we are manipulating values relative to partitions, and we could use it to encapsulate the SmallVec<T; 4> optimization. What do you think?